blocks/graph.py

```diff
@@ -520,3 +521,82 @@ def apply_dropout(computation_graph, variables, drop_prob, rng=None,
         replacement.tag.replacement_of = variable
 
     return computation_graph.replace(replacements)
+
+
+def apply_batch_normalization(computation_graph, variables, gammas,
+                              betas, axis=0, epsilon=1e-7):
+    """Returns a graph to variables in a computational graph.
+
+    Parameters
+    ----------
+    computation_graph : instance of :class:`ComputationGraph`
+        The computation graph.
+    variables : list of :class:`~tensor.TensorVariable`
```

Comment:
> For fully-connected networks what I understand is that they batch-normalize post-nonlinearity activations (this is what I read as "activations", but I might be wrong).

I agree that it's not clear what they mean by that in the paper, but I reached your same conclusions basing on:
* "The scaled and shifted values _y_ are passed to other network layers." (page 4, top left, where _y_ is the output of the BN) 
* "Any layer that previously received _x_ as the input, now receives BN(_x_)" (page 4, bottom left)
* Point 4 of algorithm 2

Nonetheless, the argument they make to justify the use of the transformation before the non-linearity in the convolutional networks section seems to be valid in the fully-connected case too (since applying BN to the post-nonlinearity activation would the same as applying it to the input of the next layer, if I am not confused):
> We add the BN transform immediately before the nonlinearity, by normalizing x = W u + b. We could have also normalized the layer inputs u, but since u is likely the output of another nonlinearity, the shape of its distribution is likely to change during training, and constraining its first and second moments would not eliminate the covariate shift.